src/google/protobuf/compiler/objectivec/method_dump.sh

```diff
@@ -0,0 +1,154 @@
+#     the "proto field" name, so "setValuesForKeysWithDictionary:" will become
+#     "valuesForKeysWithDictionary".
+
+set -eu
+
+objc_code=$(cat <<'END_CODE'
+#import <Foundation/Foundation.h>
+#import <objc/runtime.h>
+
+int main(int argc, const char * argv[]) {
+  @autoreleasepool {
+    Class cls = [NSObject class];
+
+    // List out the protocols on NSObject just so we are aware if they change.
+    unsigned int protocolCount;
+    __unsafe_unretained Protocol **protocols =
+        class_copyProtocolList(cls, &protocolCount);
+    for (unsigned int i = 0; i < protocolCount; i++) {
+      printf("// Protocol: %s\n", protocol_getName(protocols[i]));
+    }
+    free(protocols);
+
+    // Grab all the properties.
+    unsigned int propCount;
+    objc_property_t *props = class_copyPropertyList(cls, &propCount);
+    NSMutableSet *reservedNames = [[NSMutableSet alloc] init];
+    for (unsigned int i = 0; i < propCount; ++i) {
+      NSString *propertyName = [NSString stringWithUTF8String:property_getName(props[i])];
+      [reservedNames addObject:propertyName];
+    }
+    free(props);
+
+    // Note that methods have 2 defaults args (_cmd and SEL) so a method "0 arg method"
+    // actually has 2.
+    unsigned int methodCount;
+    Method *methods = class_copyMethodList(cls, &methodCount);
+    for (unsigned int i = 0; i < methodCount; ++i) {
+      int argCount = method_getNumberOfArguments(methods[i]);
+      NSString *methodName =
+          [NSString stringWithUTF8String:sel_getName(method_getName(methods[i]))];
+      if (argCount == 2) {
+        [reservedNames addObject:methodName];
+      }
+      if (argCount == 3 && [methodName hasPrefix:@"set"] && methodName.length > 4) {
+        NSString *firstLetter = [methodName substringWithRange:NSMakeRange(3,1)];
+        firstLetter = [firstLetter lowercaseString];
+        // -5 because 3 for set, 1 for the firstLetter and 1 for the colon on the end.
+        NSString *restOfString =
+            [methodName substringWithRange:NSMakeRange(4, methodName.length - 5)];
+        methodName = [firstLetter stringByAppendingString:restOfString];
+        [reservedNames addObject:methodName];
+      }
+    }
+    free(methods);
+
+    SEL sortSelector = @selector(caseInsensitiveCompare:);
+    NSArray *array = [reservedNames.allObjects sortedArrayUsingSelector:sortSelector];
+    for (NSString *item in array) {
+      // Some items with _ in them get returned in quotes, so do not add more.
+      if ([item hasPrefix:@"\""]) {
+        printf("%s,\n", item.UTF8String);
+      } else {
+        printf("\"%s\",\n", item.UTF8String);
+      }
+    }
+  }
+  return 0;
+}
+END_CODE
+)
+
+file_header=$(cat <<'END_HEADER'
+// NSObject methods
+// Autogenerated by method_dump.sh. Do not edit by hand.
+// Date: %DATE%
+// macOS: %MACOS%
+// iOS: %IOS%
+
+END_HEADER
+)
+
+# Check to make sure we are updating the correct file.
+if ! [ -e objectivec_nsobject_methods.h ]; then
```

Comment:
using bash so `[[` and `]]` instead of `[` and `]`.  Also move the `!` inside the test then.

Reply:
Done